service/src/MSRIOGroup.cpp

```diff
@@ -328,11 +332,107 @@ namespace geopm
                                                    IOGroup::M_UNITS_WATTS,
                                                    agg_function(msr_name),
                                                    ps.description + "\n    alias_for: " + ps.msr_name + " rate of change",
-                                                   IOGroup::M_SIGNAL_BEHAVIOR_VARIABLE};
+                                                   IOGroup::M_SIGNAL_BEHAVIOR_VARIABLE,
+                                                   string_format_double};
             }
         }
     }
 
+    void MSRIOGroup::register_pcnt_scalability_signals(void)
+    {
+        // register time signal; domain board
+        std::string time_name = "MSR::TIME";
+        std::shared_ptr<Signal> time_sig = std::make_shared<TimeSignal>(m_time_zero, m_time_batch);
+        m_signal_available[time_name] = {std::vector<std::shared_ptr<Signal> >({time_sig}),
+                                         GEOPM_DOMAIN_CPU,
+                                         IOGroup::M_UNITS_SECONDS,
+                                         Agg::select_first,
+                                         "Time in seconds used to calculate pcnt_rate, acnt_rate",
+                                         IOGroup::M_SIGNAL_BEHAVIOR_VARIABLE,
+                                         string_format_double};
+        int derivative_window = 8;
+        double sleep_time = 0.005;  // 5000 us
+
+        // Mapping of high-level signal name to description and
+        // underlying energy MSR.  The domain will match that of the
+        // energy signal.
+        struct cnt_data
+        {
+            std::string cnt_name;
+            std::string description;
+            std::string msr_name;
+        };
+        std::vector<cnt_data> cnt_signals {
+            {"MSR::PCNT_RATE",
+                    "Average cpu pcnt rate over 8 control loop iterations (40ms if using geopmread)",
+                    "MSR::PPERF:PCNT"},
+            {"MSR::ACNT_RATE",
+                    "Average cpu acnt rate over 8 control loop iterations (40ms if using geopmread)",
+                    "MSR::APERF:ACNT"}
```

Comment:
MSR::PCNT_RATE may give the impression that an actual MSR provides this data, when it's really a derivative of the MSR::PPERF:PCNT signal.  PCNT_RATE may be a better choice?

Reply:
It may be reasonable to use CPU_COMPUTE_ACTIVITY, to indicate this is similar to GPU_COMPUTE_ACTIVITY

Reply:
Do PCNT-per-second and ACNT-per-second have any use meaning other than their ratio? Maybe these could be excluded from the public signal names, and just give their unnamed shared pointers to the named divisionsignal?

Reply:
They may be used to build additional division signals (such as ACNT/MCNT for actual frequency), but on their own they don't add much value.  

These were initially included to provide visibility to the end user into what generated the CPU_SCALABILITY_RATIO, but if that is not required the *_RATE signals can be hidden as suggested

Reply:
Removed intermediate signals from signals list in latest.